File: src/test/java/learnyouakotlin/end/java/SessionTests.java
package learnyouakotlin.end.java;

import org.junit.Test;

import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;

public class SessionTests {
    SessionCode code = SessionCode.of(1);

    Session original = new Session(code, "The Title", null, new Presenter("Alice"));

    @Test
    public void can_change_presenters() {
        assertThat(original.withPresenters(asList(new Presenter("Bob"), new Presenter("Carol"))), equalTo(
                new Session(code, "The Title", null, new Presenter("Bob"), new Presenter("Carol"))));
    }

    @Test
    public void can_change_title() {
        assertThat(original.withTitle("Another Title"), equalTo(
                new Session(code, "Another Title", null, new Presenter("Alice"))));
    }

    @Test
    public void can_change_subtitle() {
        assertThat(original.withSubitle("The Subtitle"), equalTo(
                new Session(code, "The Title", "The Subtitle", new Presenter("Alice"))));
    }

    @Test
    public void can_remove_subtitle() {
        assertThat(original.withSubitle("The Subtitle").withSubitle(null), equalTo(
                new Session(code, "The Title", null, new Presenter("Alice"))));
    }
}

File: sample-solution/learnyouakotlin/solution/SessionTests.kt
package learnyouakotlin.solution

import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import java.util.Arrays.asList

class SessionTests {
    val original = Session("The Title", null, 2..3, Presenter("Alice"))
    
    @Test
    fun can_change_presenters() {
        assertThat(original.withPresenters(asList(Presenter("Bob"), Presenter("Carol"))), equalTo(
            Session("The Title", null, 2..3, Presenter("Bob"), Presenter("Carol"))))
    }
    
    @Test
    fun can_change_title() {
        assertThat(original.withTitle("Another Title"), equalTo(
            Session("Another Title", null, 2..3, Presenter("Alice"))))
    }
    
    @Test
    fun can_change_subtitle() {
        assertThat(original.withSubtitle("The Subtitle"), equalTo(
            Session("The Title", "The Subtitle", 2..3, Presenter("Alice"))))
    }
    
    @Test
    fun can_remove_subtitle() {
        assertThat(original.withSubtitle("The Subtitle").withSubtitle(null), equalTo(
            Session("The Title", null, 2..3, Presenter("Alice"))))
    }
}

